src/IonParserTextRaw.ts

```diff
@@ -1009,116 +1008,141 @@ private _test_symbol_as_annotation() : boolean {
-                      break;
-                  case CH_SQ:
-                      if(this.verifyTriple(index)){
-                          index = this._skip_triple_quote_gap(index, this._end, acceptComments);
-                      } else {
-                          s += String.fromCharCode(ch);
-                      }
-                      break;
-
-                  default:
-                      s += String.fromCharCode(ch);
-                      break;
-              }
-          }
-          break;
-      case T_STRING3:
-          acceptComments = true;
-          for(index = this._start; index < this._end; index++) {
-              ch = this._in.valueAt(index);
-              switch (ch) {
-                  case CH_BS:
-                      escaped = this._read_escape_sequence(index, this._end);
-                      if(escaped >= 0){
-                          s += String.fromCharCode(escaped);
-                      }
-                      index += this._esc_len;
-                      break;
-                  case CH_SQ:
-                      if(this.verifyTriple(index)){
-                          index = this._skip_triple_quote_gap(index, this._end, acceptComments);
-                      } else {
-                          s += String.fromCharCode(ch);
-                      }
-                      break;
-
-                  default:
-                      s += String.fromCharCode(ch);
-                      break;
-              }
-          }
-        break;
-      default:
-        this._error("can't get this value as a string");
-        break;
+    private isLowSurrogate(ch : number) : boolean{
+        return ch > 0xDC00 && ch < 0xDFFF;
+    }
+
+    get_value_as_string(t: number) : string {
+        let index : number;
+        let ch : number;
+        let escaped : number;
+        let acceptComments : boolean;
+        let s : string = "";
+        switch (t) {
+            case T_NULL:
+            case T_BOOL:
+            case T_INT:
+            case T_HEXINT:
+            case T_FLOAT:
+            case T_FLOAT_SPECIAL:
+            case T_DECIMAL:
+            case T_TIMESTAMP:
+            case T_IDENTIFIER:
+            case T_OPERATOR:
+            case T_BLOB:
+                for (index = this._start; index < this._end; index++) {
+                    s += String.fromCharCode(this._in.valueAt(index));
+                }
+                break;
+            case T_STRING1:
+            case T_STRING2:
+                for (index = this._start; index < this._end; index++) {
+                    ch = this._in.valueAt(index);
+                    if (ch == CH_BS) {
+                        ch = this._read_escape_sequence(index, this._end);
+                        index += this._esc_len;
+                    }
+                    if(this.isHighSurrogate(ch)){
+                        index++;
+                        let tempChar = this._in.valueAt(index);
```

Comment:
Won't this still potentially go out of bounds? `index++` on the previous line without checking if it went past the end.

Reply:
Out of bounds returns NaN(valueAt() calls charAt() underneath the hood) which will make the isLowSurrogate throw("illegal high surrogate") seemed like an elegant solution.

Reply:
I see.